{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 314,
   "id": "ee9ecd2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybiomart import Dataset\n",
    "from pandas_plink import read_plink\n",
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "id": "979a578e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reactome\n",
      "reactome_gene\n",
      "reactome_transcript\n"
     ]
    }
   ],
   "source": [
    "dataset = Dataset(name='btaurus_gene_ensembl',\n",
    "                  host='http://www.ensembl.org')\n",
    "\n",
    "for attribute in dataset.attributes.keys():\n",
    "    if 'reactome' in attribute:\n",
    "        print(attribute)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "cf8eb267",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = dataset.query(attributes=[\n",
    "    'ensembl_gene_id',\n",
    "    'external_gene_name',\n",
    "    'go_id',\n",
    "    'reactome'\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "d3a933a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:01<00:00,  2.27it/s]\n"
     ]
    }
   ],
   "source": [
    "gen_path = '../data/ML/BBB2023_MD'\n",
    "bim, fam, bed = read_plink(gen_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "id": "6ce7a4ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_123646/3430236937.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  snps_pos.rename(columns={'snp':'title', 'pos': 'End', 'chrom': 'Chromosome'}, inplace=True)\n",
      "/tmp/ipykernel_123646/3430236937.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  snps_pos['Start'] = snps_pos['End']-1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>End</th>\n",
       "      <th>Start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BTA-101873</td>\n",
       "      <td>1</td>\n",
       "      <td>776231</td>\n",
       "      <td>776230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARS-BFGL-NGS-16466</td>\n",
       "      <td>1</td>\n",
       "      <td>907810</td>\n",
       "      <td>907809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARS-BFGL-NGS-105096</td>\n",
       "      <td>1</td>\n",
       "      <td>993060</td>\n",
       "      <td>993059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BES1_Contig627_1906</td>\n",
       "      <td>1</td>\n",
       "      <td>1032564</td>\n",
       "      <td>1032563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARS-BFGL-NGS-98142</td>\n",
       "      <td>1</td>\n",
       "      <td>1110393</td>\n",
       "      <td>1110392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36299</th>\n",
       "      <td>ARS-BFGL-NGS-16401</td>\n",
       "      <td>29</td>\n",
       "      <td>50416855</td>\n",
       "      <td>50416854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36300</th>\n",
       "      <td>ARS-BFGL-NGS-3284</td>\n",
       "      <td>29</td>\n",
       "      <td>50522805</td>\n",
       "      <td>50522804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36301</th>\n",
       "      <td>BTA-66351</td>\n",
       "      <td>29</td>\n",
       "      <td>50548328</td>\n",
       "      <td>50548327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36302</th>\n",
       "      <td>ARS-BFGL-NGS-91027</td>\n",
       "      <td>29</td>\n",
       "      <td>50595638</td>\n",
       "      <td>50595637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36303</th>\n",
       "      <td>ARS-BFGL-NGS-75820</td>\n",
       "      <td>29</td>\n",
       "      <td>50621629</td>\n",
       "      <td>50621628</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36304 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+---------------------+--------------+-----------+-----------+\n",
       "| title               | Chromosome   | End       | Start     |\n",
       "| (string)            | (category)   | (int64)   | (int64)   |\n",
       "|---------------------+--------------+-----------+-----------|\n",
       "| BTA-101873          | 1            | 776231    | 776230    |\n",
       "| ARS-BFGL-NGS-16466  | 1            | 907810    | 907809    |\n",
       "| ARS-BFGL-NGS-105096 | 1            | 993060    | 993059    |\n",
       "| BES1_Contig627_1906 | 1            | 1032564   | 1032563   |\n",
       "| ...                 | ...          | ...       | ...       |\n",
       "| ARS-BFGL-NGS-3284   | 29           | 50522805  | 50522804  |\n",
       "| BTA-66351           | 29           | 50548328  | 50548327  |\n",
       "| ARS-BFGL-NGS-91027  | 29           | 50595638  | 50595637  |\n",
       "| ARS-BFGL-NGS-75820  | 29           | 50621629  | 50621628  |\n",
       "+---------------------+--------------+-----------+-----------+\n",
       "Unstranded PyRanges object has 36,304 rows and 4 columns from 29 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_pos = bim[['snp','chrom', 'pos']]\n",
    "# counts = snps_pos.groupby('snp').cumcount()\n",
    "# snps_pos.loc[counts > 0, 'snp'] = snps_pos.loc[counts > 0, 'snp'] + '_' + counts.loc[counts > 0].astype(str)\n",
    "\n",
    "snps_pos.rename(columns={'snp':'title', 'pos': 'End', 'chrom': 'Chromosome'}, inplace=True)\n",
    "snps_pos['Start'] = snps_pos['End']-1\n",
    "\n",
    "# snps_gr = pr.PyRanges(\n",
    "#     chromosomes = snps_pos[\"chrom\"],\n",
    "#     starts = snps_pos[\"pos\"] -1,\n",
    "#     ends=snps_pos[\"pos\"],\n",
    "# )\n",
    "\n",
    "snps_gr = pr.PyRanges(snps_pos)\n",
    "snps_gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d09120a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pr.read_gtf(\"../data/Bos_taurus.UMD3.1.90.gtf\")\n",
    "\n",
    "genes = genes[genes.Feature == \"gene\"]\n",
    "\n",
    "overlap = snps_gr.join(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd5a597c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['title', 'Chromosome', 'End', 'Start', 'Source', 'Feature', 'Start_b',\n",
      "       'End_b', 'Score', 'Strand', 'Frame', 'gene_id', 'gene_version',\n",
      "       'gene_name', 'gene_source', 'gene_biotype'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "overlap = overlap[['title','Chromosome', 'Start', 'End', 'Source', 'Feature', 'Start_b', 'End_b',\n",
    "       'Score', 'Strand', 'Frame', 'gene_id', 'gene_version', 'gene_name',\n",
    "       'gene_source', 'gene_biotype']]\n",
    "\n",
    "print(overlap.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7178a762",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>End</th>\n",
       "      <th>Start</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARS-BFGL-NGS-66449</td>\n",
       "      <td>1</td>\n",
       "      <td>1204825</td>\n",
       "      <td>1204824</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>1189227</td>\n",
       "      <td>1222846</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSBTAG00000008490</td>\n",
       "      <td>5</td>\n",
       "      <td>CRYZL1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hapmap53766-ss46526150</td>\n",
       "      <td>1</td>\n",
       "      <td>2082139</td>\n",
       "      <td>2082138</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>2067035</td>\n",
       "      <td>2099875</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSBTAG00000003064</td>\n",
       "      <td>5</td>\n",
       "      <td>PAXBP1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARS-BFGL-NGS-62826</td>\n",
       "      <td>1</td>\n",
       "      <td>2158691</td>\n",
       "      <td>2158690</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>2109218</td>\n",
       "      <td>2196114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSBTAG00000003063</td>\n",
       "      <td>5</td>\n",
       "      <td>SYNJ1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARS-BFGL-NGS-39992</td>\n",
       "      <td>1</td>\n",
       "      <td>2391777</td>\n",
       "      <td>2391776</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>2382872</td>\n",
       "      <td>2459950</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSBTAG00000012412</td>\n",
       "      <td>4</td>\n",
       "      <td>URB1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BTB-00001612</td>\n",
       "      <td>1</td>\n",
       "      <td>3492221</td>\n",
       "      <td>3492220</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>3483532</td>\n",
       "      <td>3721930</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSBTAG00000017839</td>\n",
       "      <td>5</td>\n",
       "      <td>TIAM1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    title Chromosome      End    Start   Source Feature  \\\n",
       "0      ARS-BFGL-NGS-66449          1  1204825  1204824  ensembl    gene   \n",
       "1  Hapmap53766-ss46526150          1  2082139  2082138  ensembl    gene   \n",
       "2      ARS-BFGL-NGS-62826          1  2158691  2158690  ensembl    gene   \n",
       "3      ARS-BFGL-NGS-39992          1  2391777  2391776  ensembl    gene   \n",
       "4            BTB-00001612          1  3492221  3492220  ensembl    gene   \n",
       "\n",
       "   Start_b    End_b Score Strand Frame             gene_id gene_version  \\\n",
       "0  1189227  1222846     .      +     .  ENSBTAG00000008490            5   \n",
       "1  2067035  2099875     .      +     .  ENSBTAG00000003064            5   \n",
       "2  2109218  2196114     .      +     .  ENSBTAG00000003063            5   \n",
       "3  2382872  2459950     .      +     .  ENSBTAG00000012412            4   \n",
       "4  3483532  3721930     .      +     .  ENSBTAG00000017839            5   \n",
       "\n",
       "  gene_name gene_source    gene_biotype  \n",
       "0    CRYZL1     ensembl  protein_coding  \n",
       "1    PAXBP1     ensembl  protein_coding  \n",
       "2     SYNJ1     ensembl  protein_coding  \n",
       "3      URB1     ensembl  protein_coding  \n",
       "4     TIAM1     ensembl  protein_coding  "
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap.df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61f8588b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  source      native                                name       p_value  \\\n",
      "0  GO:BP  GO:0065007               biological regulation  7.361622e-80   \n",
      "1  GO:BP  GO:0050789    regulation of biological process  3.232581e-71   \n",
      "2  GO:BP  GO:0048856    anatomical structure development  2.762056e-69   \n",
      "3  GO:BP  GO:0032502               developmental process  9.587048e-67   \n",
      "4  GO:BP  GO:0007275  multicellular organism development  2.003839e-60   \n",
      "\n",
      "   significant                                        description  term_size  \\\n",
      "0         True  \"Any process that modulates a measurable attri...       8335   \n",
      "1         True  \"Any process that modulates the frequency, rat...       8113   \n",
      "2         True  \"The biological process whose specific outcome...       3832   \n",
      "3         True  \"A biological process whose specific outcome i...       4098   \n",
      "4         True  \"The biological process whose specific outcome...       3013   \n",
      "\n",
      "   query_size  intersection_size  effective_domain_size  precision    recall  \\\n",
      "0        3683               2532                  15111   0.687483  0.303779   \n",
      "1        3683               2453                  15111   0.666033  0.302354   \n",
      "2        3683               1358                  15111   0.368721  0.354384   \n",
      "3        3683               1423                  15111   0.386370  0.347243   \n",
      "4        3683               1101                  15111   0.298941  0.365417   \n",
      "\n",
      "     query                   parents  \\\n",
      "0  query_1              [GO:0008150]   \n",
      "1  query_1  [GO:0008150, GO:0065007]   \n",
      "2  query_1              [GO:0032502]   \n",
      "3  query_1              [GO:0008150]   \n",
      "4  query_1  [GO:0032501, GO:0048856]   \n",
      "\n",
      "                                       intersections  \\\n",
      "0  [PAXBP1, SYNJ1, TIAM1, APP, SAMSN1, EPHA3, ARL...   \n",
      "1  [PAXBP1, SYNJ1, TIAM1, APP, SAMSN1, EPHA3, ARL...   \n",
      "2  [TIAM1, APP, EPHA3, ARL13B, COL8A1, TBC1D23, A...   \n",
      "3  [TIAM1, APP, EPHA3, ARL13B, COL8A1, TBC1D23, A...   \n",
      "4  [APP, EPHA3, ARL13B, COL8A1, TBC1D23, ALCAM, B...   \n",
      "\n",
      "                                           evidences  \n",
      "0  [[IBA, IEA], [IEA], [IDA, IEA], [IEA], [IBA], ...  \n",
      "1  [[IBA, IEA], [IEA], [IDA, IEA], [IEA], [IBA], ...  \n",
      "2  [[IDA], [IBA, IEA], [IEA], [IEA], [IEA], [IMP]...  \n",
      "3  [[IDA], [IBA, IEA], [IEA], [IEA], [IEA], [IMP]...  \n",
      "4  [[IBA, IEA], [IEA], [IEA], [IEA], [IMP], [IEA]...  \n"
     ]
    }
   ],
   "source": [
    "from gprofiler.gprofiler import GProfiler\n",
    "\n",
    "gp = GProfiler(return_dataframe=True)\n",
    "\n",
    "# List of unique gene symbols\n",
    "gene_list = overlap.df['gene_name'].dropna().unique().tolist()\n",
    "\n",
    "# Query for pathways\n",
    "gp = GProfiler(return_dataframe=True)\n",
    "\n",
    "results = gp.profile(\n",
    "    organism='btaurus',\n",
    "    query=gene_list,\n",
    "    sources=['KEGG', 'GO:BP', 'REAC'],  # KEGG, GO biological process, Reactome\n",
    "    no_evidences=False\n",
    ")\n",
    "\n",
    "# View results\n",
    "print(results.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "462291d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      GO:0065007\n",
       "1      GO:0050789\n",
       "2      GO:0048856\n",
       "3      GO:0032502\n",
       "4      GO:0007275\n",
       "          ...    \n",
       "586    GO:0021915\n",
       "587    GO:0021954\n",
       "588    GO:1903131\n",
       "589    GO:0007611\n",
       "590    GO:0090130\n",
       "Name: native, Length: 591, dtype: object"
      ]
     },
     "execution_count": 261,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results['native']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45a83bcd",
   "metadata": {},
   "source": [
    "### Constructing the neural network using information from the genes and pathways on the structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b01ea6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "counts = overlap.df.groupby('title').cumcount()\n",
    "overlap.df['title'] = overlap.df['title'] + '_' +counts.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ff0b6f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# snp_list = overlap.df['title'].tolist()\n",
    "# gene_list = overlap.df['gene_name'].unique().tolist()\n",
    "\n",
    "# n_snps = len(set(snp_list))\n",
    "# n_genes = len(gene_list)\n",
    "\n",
    "# # Map SNP/Gene to index\n",
    "# snp_idx = {}\n",
    "# for i, snp in enumerate(snp_list):\n",
    "#     if snp in snp_idx.keys():\n",
    "#         snp_idx[snp] = snp_idx[snp]+[i]\n",
    "#     else:\n",
    "#         snp_idx[snp] = [i]\n",
    "\n",
    "# gene_idx = {gene:i for i,gene in enumerate(gene_list)}\n",
    "\n",
    "# # Initialize mask\n",
    "# mask_snp_gene = np.zeros((n_genes, n_snps), dtype=np.float32)\n",
    "\n",
    "# for _, row in overlap.df.iterrows():\n",
    "#     i = gene_idx[row['gene_name']]\n",
    "#     for j in snp_idx[row['title']]:\n",
    "#         mask_snp_gene[i,j] = 1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87463eff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The snp_genes_unique has shape (11208, 2)\n",
      "The snp_list has 11208 size\n",
      "The n_snps has 11208 size\n",
      "The snp_idx has 11208\n",
      "The mask_snp_gene has shape (6055, 11208)\n",
      "The sum of the mask_snp_gene is 11295.0\n"
     ]
    }
   ],
   "source": [
    "snp_genes = overlap.df[['title', 'gene_id']]\n",
    "\n",
    "snp_genes_unique = (\n",
    "    snp_genes\n",
    "    .groupby('title', as_index=False)['gene_id']\n",
    "    .agg(list)\n",
    ")\n",
    "\n",
    "print(f\"The snp_genes_unique has shape {snp_genes_unique.shape}\")\n",
    "\n",
    "snp_list = snp_genes_unique['title'].tolist()\n",
    "gene_list = overlap.df['gene_id'].unique().tolist()\n",
    "\n",
    "print(f\"The snp_list has {len(snp_list)} size\")\n",
    "\n",
    "n_snps = len(snp_list)\n",
    "n_genes = len(gene_list)\n",
    "\n",
    "print(f\"The n_snps has {n_snps} size\")\n",
    "\n",
    "snp_idx = {snp:i for i,snp in enumerate(snp_list)}\n",
    "gene_idx = {gene:i for i,gene in enumerate(gene_list)}\n",
    "\n",
    "print(f\"The snp_idx has {len(snp_idx)}\")\n",
    "\n",
    "mask_snp_gene = np.zeros((n_genes, n_snps), dtype=np.float32)\n",
    "\n",
    "for snp in snp_list:\n",
    "    j = snp_idx[snp]\n",
    "    for gene in snp_genes_unique.loc[snp_genes_unique['title'] == snp,'gene_id'].iloc[0]:\n",
    "        i = gene_idx[gene]\n",
    "        mask_snp_gene[i,j] = 1.0\n",
    "\n",
    "print(f\"The mask_snp_gene has shape {mask_snp_gene.shape}\")\n",
    "\n",
    "print(f\"The sum of the mask_snp_gene is {mask_snp_gene.sum()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6d694c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "288 .... 2.0\n",
      "473 .... 2.0\n",
      "489 .... 2.0\n",
      "511 .... 2.0\n",
      "597 .... 2.0\n",
      "627 .... 2.0\n",
      "871 .... 2.0\n",
      "969 .... 2.0\n",
      "1163 .... 2.0\n",
      "1217 .... 2.0\n",
      "1234 .... 2.0\n",
      "1411 .... 2.0\n",
      "1682 .... 2.0\n",
      "1738 .... 3.0\n",
      "1744 .... 2.0\n",
      "1813 .... 2.0\n",
      "1995 .... 2.0\n",
      "2064 .... 2.0\n",
      "2189 .... 2.0\n",
      "2420 .... 2.0\n",
      "2466 .... 2.0\n",
      "2527 .... 2.0\n",
      "2746 .... 2.0\n",
      "2754 .... 2.0\n",
      "2826 .... 2.0\n",
      "2936 .... 2.0\n",
      "2998 .... 2.0\n",
      "3060 .... 2.0\n",
      "3233 .... 2.0\n",
      "3371 .... 2.0\n",
      "3499 .... 2.0\n",
      "3703 .... 2.0\n",
      "3790 .... 2.0\n",
      "3939 .... 2.0\n",
      "4192 .... 2.0\n",
      "4515 .... 2.0\n",
      "4538 .... 2.0\n",
      "4596 .... 2.0\n",
      "4604 .... 2.0\n",
      "4808 .... 2.0\n",
      "4891 .... 2.0\n",
      "5289 .... 2.0\n",
      "5330 .... 2.0\n",
      "5805 .... 2.0\n",
      "5906 .... 2.0\n",
      "6081 .... 2.0\n",
      "6362 .... 2.0\n",
      "6488 .... 2.0\n",
      "6531 .... 2.0\n",
      "6573 .... 2.0\n",
      "6612 .... 2.0\n",
      "6899 .... 2.0\n",
      "7025 .... 2.0\n",
      "7089 .... 2.0\n",
      "7151 .... 2.0\n",
      "7222 .... 2.0\n",
      "7769 .... 2.0\n",
      "7956 .... 2.0\n",
      "8037 .... 2.0\n",
      "8048 .... 2.0\n",
      "8126 .... 2.0\n",
      "8449 .... 2.0\n",
      "8616 .... 2.0\n",
      "8823 .... 2.0\n",
      "8871 .... 2.0\n",
      "8880 .... 2.0\n",
      "8893 .... 2.0\n",
      "8894 .... 2.0\n",
      "9013 .... 2.0\n",
      "9014 .... 2.0\n",
      "9293 .... 2.0\n",
      "9300 .... 2.0\n",
      "9811 .... 2.0\n",
      "9861 .... 2.0\n",
      "9924 .... 2.0\n",
      "9960 .... 2.0\n",
      "10013 .... 2.0\n",
      "10030 .... 2.0\n",
      "10300 .... 2.0\n",
      "10385 .... 2.0\n",
      "10456 .... 2.0\n",
      "10471 .... 2.0\n",
      "10568 .... 2.0\n",
      "10609 .... 2.0\n",
      "10755 .... 2.0\n",
      "11176 .... 2.0\n"
     ]
    }
   ],
   "source": [
    "# for i,_ in enumerate(mask_snp_gene.T):\n",
    "#     if mask_snp_gene.T[i,:].sum() >1 :\n",
    "#         print(f\"{i} .... {mask_snp_gene.T[i,:].sum()}\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c936e7f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARS-BFGL-NGS-101891\n"
     ]
    }
   ],
   "source": [
    "# for snp in snp_idx.keys():\n",
    "#     if snp_idx[snp] == 473:\n",
    "#         print(snp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f724924f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>ARS-BFGL-NGS-101891</td>\n",
       "      <td>[ENSBTAG00000010577, ENSBTAG00000010571]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   title                                   gene_id\n",
       "473  ARS-BFGL-NGS-101891  [ENSBTAG00000010577, ENSBTAG00000010571]"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# snp_genes_unique[snp_genes_unique['title'] == 'ARS-BFGL-NGS-101891']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f097d8a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSBTAG00000017753'"
      ]
     },
     "execution_count": 271,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap.df.loc[ overlap.df['gene_name'] == \"APP\", 'gene_id'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4077254",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = overlap.df[['gene_name', 'gene_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dde325a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Defining pathways\n",
    "# pathway_list = results['native'].unique()\n",
    "\n",
    "# n_pathway = len(pathway_list)\n",
    "\n",
    "# # Mapping pathway to index\n",
    "# idx_pathway = {pathway:i for i,pathway in enumerate(pathway_list)}\n",
    "\n",
    "# mask_gene_pathway = np.zeros((n_pathway, n_genes), dtype=np.float32)\n",
    "\n",
    "# for _,row in results.iterrows():\n",
    "#     genes_intersec = row['intersections']\n",
    "#     i = idx_pathway[row['native']]\n",
    "#     for gene in genes_intersec:\n",
    "#         gene_id = gene_names[ gene_names['gene_name'] == gene, 'gene_id'].iloc[0]\n",
    "#         j = gene_idx[gene_id]\n",
    "#         mask_gene_pathway[i,j] = 1.0\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e801b8cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining pathways\n",
    "pathway_list = results['native'].unique()\n",
    "n_pathway = len(pathway_list)\n",
    "\n",
    "idx_pathway = {pathway: i for i, pathway in enumerate(pathway_list)}\n",
    "\n",
    "gene_name_to_id = dict(zip(gene_names['gene_name'], gene_names['gene_id']))\n",
    "\n",
    "mask_gene_pathway = np.zeros((n_pathway, n_genes), dtype=np.float32)\n",
    "\n",
    "for row in results.itertuples(index=False):\n",
    "    i = idx_pathway[row.native]\n",
    "\n",
    "    for gene in row.intersections:\n",
    "        gene_id = gene_name_to_id.get(gene)\n",
    "        if gene_id is None:\n",
    "            continue\n",
    "        \n",
    "        j = gene_idx.get(gene_id)\n",
    "        if j is not None:\n",
    "            mask_gene_pathway[i, j] = 1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2db9504",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(591, 6055)\n",
      "135740.0\n"
     ]
    }
   ],
   "source": [
    "print(mask_gene_pathway.shape)\n",
    "print(mask_gene_pathway.sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4757f234",
   "metadata": {},
   "outputs": [],
   "source": [
    "class MaskedLinear(nn.Module):\n",
    "    def __init__(self, in_features, out_features, mask):\n",
    "        super().__init__()\n",
    "\n",
    "        self.weight = nn.Parameter(torch.randn(out_features, in_features)*0.01)\n",
    "        self.bias = nn.Parameter(torch.zeros(out_features))\n",
    "\n",
    "        self.register_buffer(\"mask\", mask)\n",
    "\n",
    "    def forward(self, x):\n",
    "        masked_weight = self.weight * self.mask\n",
    "        return F.linear(x, masked_weight, self.bias)\n",
    "    \n",
    "class PartialNet(nn.Module):\n",
    "    def __init__(self, input_dim, hidden1, hidden2, fc1_dim, fc2_dim, mask1, mask2):\n",
    "        super().__init__()\n",
    "\n",
    "        # Masked layers\n",
    "        self.masked1 = MaskedLinear(input_dim, hidden1, mask1)\n",
    "        self.masked2 = MaskedLinear(hidden1, hidden2, mask2)\n",
    "\n",
    "        # FC layers\n",
    "        self.fc1 = nn.Linear(hidden2, fc1_dim)\n",
    "        self.fc2 = nn.Linear(fc1_dim, fc2_dim)\n",
    "        self.fc3 = nn.Linear(fc2_dim, 1)\n",
    "\n",
    "        # Activation\n",
    "        self.activation = nn.ReLU()\n",
    "\n",
    "    def forward(self, x):\n",
    "        x = self.activation(self.masked1(x))\n",
    "        x = self.activation(self.masked2(x))\n",
    "        x = self.activation(self.fc1(x))\n",
    "        x = self.activation(self.fc2(x))\n",
    "        x = self.fc3(x)\n",
    "        return x\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78304e05",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"None of ['snp'] are in the columns\"",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[32m/tmp/ipykernel_123646/1430047907.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m()\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Selectioning the correct SNPs\u001b[39;00m\n\u001b[32m      2\u001b[39m \n\u001b[32m----> \u001b[39m\u001b[32m3\u001b[39m bim = bim.set_index(\u001b[33m'snp'\u001b[39m)\n\u001b[32m      4\u001b[39m snp_used = snp_genes_unique.set_index(\u001b[33m'title'\u001b[39m)\n\u001b[32m      5\u001b[39m snp_used.index.name = \u001b[33m'snp'\u001b[39m\n\u001b[32m      6\u001b[39m snp_to_index = bim.loc[snp_used.index][\u001b[33m'i'\u001b[39m]\n",
      "\u001b[32m~/Documents/Doutorado/projects/Snp_models/bio_informed_DNN/venv39_bioDNN/lib/python3.13/site-packages/pandas/core/frame.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(self, keys, drop, append, inplace, verify_integrity)\u001b[39m\n\u001b[32m   6140\u001b[39m                     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28;01mnot\u001b[39;00m found:\n\u001b[32m   6141\u001b[39m                         missing.append(col)\n\u001b[32m   6142\u001b[39m \n\u001b[32m   6143\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m missing:\n\u001b[32m-> \u001b[39m\u001b[32m6144\u001b[39m             \u001b[38;5;28;01mraise\u001b[39;00m KeyError(f\"None of {missing} are in the columns\")\n\u001b[32m   6145\u001b[39m \n\u001b[32m   6146\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m inplace:\n\u001b[32m   6147\u001b[39m             frame = self\n",
      "\u001b[31mKeyError\u001b[39m: \"None of ['snp'] are in the columns\""
     ]
    }
   ],
   "source": [
    "# Selectioning the correct SNPs\n",
    "\n",
    "bim = bim.set_index('snp')\n",
    "snp_used = snp_genes_unique.set_index('title')\n",
    "snp_used.index.name = 'snp'\n",
    "snp_to_index = bim.loc[snp_used.index]['i']\n",
    "\n",
    "# Getting the matrix indexed correctly\n",
    "X = bed[snp_to_index]#.compute()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa804b0b",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "boolean index did not match indexed array along dimension 0; dimension is 11208 but corresponding boolean dimension is 18324",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mIndexError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[312]\u001b[39m\u001b[32m, line 9\u001b[39m\n\u001b[32m      7\u001b[39m \u001b[38;5;66;03m# Separating data to training\u001b[39;00m\n\u001b[32m      8\u001b[39m y = pheno_df[\u001b[32m2\u001b[39m][~pheno_df[\u001b[32m2\u001b[39m].isna()].to_numpy()\n\u001b[32m----> \u001b[39m\u001b[32m9\u001b[39m X = \u001b[43mX\u001b[49m\u001b[43m.\u001b[49m\u001b[43mT\u001b[49m\u001b[43m[\u001b[49m\u001b[43m~\u001b[49m\u001b[43mpheno_df\u001b[49m\u001b[43m[\u001b[49m\u001b[32;43m2\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m.\u001b[49m\u001b[43misna\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m.compute()\n",
      "\u001b[31mIndexError\u001b[39m: boolean index did not match indexed array along dimension 0; dimension is 11208 but corresponding boolean dimension is 18324"
     ]
    }
   ],
   "source": [
    "pheno_df = pd.read_csv(\"../data/ML/pheno_2023bbb_0twins_6traits_mask\", delimiter=\"\\t\", header=None)\n",
    "\n",
    "pheno_df = pheno_df.drop(columns=[0])\n",
    "pheno_df = pheno_df.set_index(1)\n",
    "pheno_df.index.name == \"iid\"\n",
    "\n",
    "# Separating data to training\n",
    "y = pheno_df[2][~pheno_df[2].isna()].to_numpy()\n",
    "X = X.T[~pheno_df[2].isna()].compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df3996ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train/test split\n",
    "X_train, X_test, y_train, y_test = train_test_split(\n",
    "    X, y, test_size=1000, random_state=42)\n",
    "\n",
    "scaler_X = StandardScaler()\n",
    "\n",
    "X_train = scaler_X.fit_transform(X_train)   # fit ONLY on train\n",
    "X_test  = scaler_X.transform(X_test) \n",
    "\n",
    "X_train = torch.tensor(X_train, dtype=torch.float32)\n",
    "X_test  = torch.tensor(X_test, dtype=torch.float32).unsqueeze(1)\n",
    "\n",
    "scaler_y = StandardScaler()\n",
    "\n",
    "y_train = scaler_y.fit_transform(y_train.reshape(-1, 1))\n",
    "y_test  = scaler_y.transform(y_test.reshape(-1, 1))\n",
    "\n",
    "y_train = torch.tensor(y_train, dtype=torch.float32).unsqueeze(1)\n",
    "y_test  = torch.tensor(y_test, dtype=torch.float32).unsqueeze(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93fc16e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13330, 11208)"
      ]
     },
     "execution_count": 287,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d094f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dim = n_snps\n",
    "hidden1 = n_genes\n",
    "hidden2 = n_pathway\n",
    "fc1_dim = 100\n",
    "fc2_dim = 100\n",
    "mask1 = torch.from_numpy(mask_snp_gene)\n",
    "mask2 = torch.from_numpy(mask_gene_pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "351b1d12",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = PartialNet(input_dim, hidden1, hidden2,\n",
    "                   fc1_dim, fc2_dim,\n",
    "                   mask1, mask2)\n",
    "\n",
    "criterion = nn.HuberLoss()\n",
    "optimizer = torch.optim.Adam(model.parameters(), lr=1e-5)\n",
    "\n",
    "num_epochs = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb632033",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 0, Loss: 3.152222\n",
      "Epoch 1, Loss: 3.152098\n",
      "Epoch 2, Loss: 3.151975\n",
      "Epoch 3, Loss: 3.151852\n",
      "Epoch 4, Loss: 3.151728\n",
      "Epoch 5, Loss: 3.151605\n",
      "Epoch 6, Loss: 3.151481\n",
      "Epoch 7, Loss: 3.151358\n",
      "Epoch 8, Loss: 3.151235\n",
      "Epoch 9, Loss: 3.151111\n",
      "Epoch 10, Loss: 3.150988\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[310]\u001b[39m\u001b[32m, line 8\u001b[39m\n\u001b[32m      5\u001b[39m outputs = model(torch.from_numpy(X_train).float())\n\u001b[32m      6\u001b[39m loss = criterion(outputs, torch.from_numpy(y_train).float().unsqueeze(\u001b[32m1\u001b[39m))\n\u001b[32m----> \u001b[39m\u001b[32m8\u001b[39m \u001b[43mloss\u001b[49m\u001b[43m.\u001b[49m\u001b[43mbackward\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m      9\u001b[39m optimizer.step()\n\u001b[32m     11\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mEpoch \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mepoch\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m, Loss: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mloss.item()\u001b[38;5;132;01m:\u001b[39;00m\u001b[33m.6f\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/Doutorado/projects/Snp_models/bio_informed_DNN/venv39_bioDNN/lib/python3.13/site-packages/torch/_tensor.py:630\u001b[39m, in \u001b[36mTensor.backward\u001b[39m\u001b[34m(self, gradient, retain_graph, create_graph, inputs)\u001b[39m\n\u001b[32m    620\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m has_torch_function_unary(\u001b[38;5;28mself\u001b[39m):\n\u001b[32m    621\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m handle_torch_function(\n\u001b[32m    622\u001b[39m         Tensor.backward,\n\u001b[32m    623\u001b[39m         (\u001b[38;5;28mself\u001b[39m,),\n\u001b[32m   (...)\u001b[39m\u001b[32m    628\u001b[39m         inputs=inputs,\n\u001b[32m    629\u001b[39m     )\n\u001b[32m--> \u001b[39m\u001b[32m630\u001b[39m \u001b[43mtorch\u001b[49m\u001b[43m.\u001b[49m\u001b[43mautograd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mbackward\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    631\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgradient\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mretain_graph\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcreate_graph\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minputs\u001b[49m\u001b[43m=\u001b[49m\u001b[43minputs\u001b[49m\n\u001b[32m    632\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/Doutorado/projects/Snp_models/bio_informed_DNN/venv39_bioDNN/lib/python3.13/site-packages/torch/autograd/__init__.py:364\u001b[39m, in \u001b[36mbackward\u001b[39m\u001b[34m(tensors, grad_tensors, retain_graph, create_graph, grad_variables, inputs)\u001b[39m\n\u001b[32m    359\u001b[39m     retain_graph = create_graph\n\u001b[32m    361\u001b[39m \u001b[38;5;66;03m# The reason we repeat the same comment below is that\u001b[39;00m\n\u001b[32m    362\u001b[39m \u001b[38;5;66;03m# some Python versions print out the first line of a multi-line function\u001b[39;00m\n\u001b[32m    363\u001b[39m \u001b[38;5;66;03m# calls in the traceback and some print out the last line\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m364\u001b[39m \u001b[43m_engine_run_backward\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    365\u001b[39m \u001b[43m    \u001b[49m\u001b[43mtensors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    366\u001b[39m \u001b[43m    \u001b[49m\u001b[43mgrad_tensors_\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    367\u001b[39m \u001b[43m    \u001b[49m\u001b[43mretain_graph\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    368\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcreate_graph\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    369\u001b[39m \u001b[43m    \u001b[49m\u001b[43minputs_tuple\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    370\u001b[39m \u001b[43m    \u001b[49m\u001b[43mallow_unreachable\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m    371\u001b[39m \u001b[43m    \u001b[49m\u001b[43maccumulate_grad\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m    372\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/Doutorado/projects/Snp_models/bio_informed_DNN/venv39_bioDNN/lib/python3.13/site-packages/torch/autograd/graph.py:865\u001b[39m, in \u001b[36m_engine_run_backward\u001b[39m\u001b[34m(t_outputs, *args, **kwargs)\u001b[39m\n\u001b[32m    863\u001b[39m     unregister_hooks = _register_logging_hooks_on_whole_graph(t_outputs)\n\u001b[32m    864\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m865\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mVariable\u001b[49m\u001b[43m.\u001b[49m\u001b[43m_execution_engine\u001b[49m\u001b[43m.\u001b[49m\u001b[43mrun_backward\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# Calls into the C++ engine to run the backward pass\u001b[39;49;00m\n\u001b[32m    866\u001b[39m \u001b[43m        \u001b[49m\u001b[43mt_outputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\n\u001b[32m    867\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# Calls into the C++ engine to run the backward pass\u001b[39;00m\n\u001b[32m    868\u001b[39m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[32m    869\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m attach_logging_hooks:\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "for epoch in range(num_epochs):\n",
    "    model.train()\n",
    "    optimizer.zero_grad()\n",
    "\n",
    "    outputs = model(torch.from_numpy(X_train).float())\n",
    "    loss = criterion(outputs, torch.from_numpy(y_train).float().unsqueeze(1))\n",
    "\n",
    "    loss.backward()\n",
    "    optimizer.step()\n",
    "\n",
    "    print(f\"Epoch {epoch}, Loss: {loss.item():.6f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv39_bioDNN",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
